### PR TITLE
Matrix/axes

### PR DESCRIPTION
**Pull request by bitwiseman**
Dependent on #332

* JENKINS issue(s):

    * Issue link or n/a
* Description:
    * ...
* Documentation changes:
    * Link to related jenkins.io PR or explanation for why doc change not needed
* Users/aliases to notify:
    * ...

### CODE REVIEW

File: pipeline-model-definition/src/test/resources/errors/parallelMatrixStagesGroupsDeepNesting.groovy [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+pipeline {
+    agent none
+    stages {
+        stage("foo") {
+            parallel {
+                stage("first-branch") {
+                    stages {
+                        stage("first") {
+                            maatrix {
```
Comment (sbeaulie) [inline pipeline-model-definition/src/test/resources/errors/parallelMatrixStagesGroupsDeepNesting.groovy]: typo? `maatrix` should be `matrix`

File: pipeline-model-definition/src/test/resources/matrix/nestedMatrixStages.groovy [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+pipeline {
+    agent none
+    stages {
+        stage("foo") {
+            matrix {
+                statges {
```
Comment (sbeaulie) [inline pipeline-model-definition/src/test/resources/matrix/nestedMatrixStages.groovy]: typo? `statges` should be `stages`

**Comment by sbeaulie** [comment]
I was trying to get the state of the 'matrix-like' functionality in Jenkins Pipelines, so I'm quite excited to see a PR. Is this the PR to follow? Have you put thoughts into how to reload/replay failed matrix axes?

File: pipeline-model-definition/src/test/resources/errors/parallelMatrixStagesGroupsDeepNesting.groovy (hunk 1 above)
Comment (bitwiseman) [reply pipeline-model-definition/src/test/resources/errors/parallelMatrixStagesGroupsDeepNesting.groovy]:
```suggestion
                            matrix {
```

**Comment by bitwiseman** [comment]
Pretty sure there are still failure here.  Once they are fixed I'll undraft this PR.

**Comment by bitwiseman** [comment]
The two failures here seem unrelated to my changes.  I'm going to merge and continue to look at them later.

**Pull request merged**